Let managers reset user's password

iagirre opened the pull request:
References
==========
This PR fullfills and closes #2500

Objectives
==========
Let the manager change the users' password in two ways:
1. Send an email to the registered email so that the user can be able to change the password by themselves.
2. Change the password using a form. The manager can write it in plain text or generate a random value. When the password is saved, a button to print is enabled.

Visual Changes
=======================
- Reset password button
![password01](https://user-images.githubusercontent.com/31625251/37462178-4853bf82-2851-11e8-95ad-35a2261393e1.png)

- User with email
![password01_1](https://user-images.githubusercontent.com/31625251/37462195-5c920be8-2851-11e8-9ce4-f40837478901.png)

- User without email
![password02](https://user-images.githubusercontent.com/31625251/37462203-6378311c-2851-11e8-9a75-e93fe9ddf1a8.png)

-Random password generator and show/hide password
![password02_2](https://user-images.githubusercontent.com/31625251/37462218-72cb32cc-2851-11e8-82ff-88a75c5e86af.gif)

- Password saved and print password page
![password03](https://user-images.githubusercontent.com/31625251/37462226-7cf3aacc-2851-11e8-847f-caaac0780f55.png)

- How password is printed
![password04](https://user-images.githubusercontent.com/31625251/37462509-7d6d277a-2852-11e8-8763-28a855a75461.png)

Notes
=====================
The "Reset password" button (next to "Change user") is placed with a float. When the PR #1363 merges with this one it should be placed next to "Edit account", as expected.

MariaCheca commented on app/assets/javascripts/managers.js.coffee:
```diff
@@ -0,0 +1,23 @@
+      $('#user_password').val(password)
+
+    $(".show-password").on "click", (event) ->
+      console.log("hey")
```
Woops

MariaCheca commented on app/assets/javascripts/managers.js.coffee:
```diff
@@ -0,0 +1,23 @@
+      if $("#user_password").is("input[type='password']")
+        $('#user_password').prop 'type', 'text'
+      else
+        $('#user_password').prop 'type', 'password'
```
If an object is going to be used more than twice it's a good practice to store it into a local variable.

MariaCheca commented on config/locales/es/management.yml:
```diff
@@ -5,13 +5,26 @@ es:
+          password: Contraseña
+          send_email: Enviar email de reseteo de contraseña
+          reset_email_send: Email enviado correctamente.
+          reseted: ¡Contraseña reseteada correctamente!
```
No exclamations, please 😬

iagirre commented:
I made the requested changes and changed the UI a little bit.

This is the new form when the user has an email:
![password05](https://user-images.githubusercontent.com/31625251/37585004-d5213b7a-2b57-11e8-9342-8e4f5584fa09.png)

And this is the form when the user doesn't have an email:
![password06](https://user-images.githubusercontent.com/31625251/37585269-d751c670-2b58-11e8-8f00-279975007efc.png)

bertocq commented on config/locales/es/management.yml:
```diff
@@ -5,13 +5,26 @@ es:
       show:
         title: Cuenta de usuario
+      edit:
+        title: 'Editar cuenta de usuario: Resetear contraseña'
```
I think `reestablecer` would be a better choice than `Resetear`. Same at "send_email", "reseted" and "reset_password" strings

bertocq commented on spec/features/management/account_spec.rb:
```diff
@@ -29,4 +29,91 @@
+
+    logout
+
+    sent_token = /.*reset_password_token=(.*)".*/.match(ActionMailer::Base.deliveries.last.body.to_s)[1]
```
I wouldn't test here that the reset password form works, that's already tested at https://github.com/AyuntamientoMadrid/consul/blob/master/spec/features/users_auth_spec.rb#L333-L353 instead the objective of this scenario is to check that the email is sent

bertocq commented on spec/features/management/account_spec.rb:
```diff
@@ -29,4 +29,91 @@
+
+    expect(page).to have_content 'Password reseted successfully'
+    expect(page).to have_css("a[href='javascript:window.print();']", text: 'Print password')
+    expect(page).to have_css("div.for-print-only", text: /Password:/, visible: false)
```
Are we really checking here that the print functionallity works? Or just that the links are present?

Maybe it would be better to use the "manually set password" scenario to be able to correctly check that when printing we're showing the same password that the manager typed

iagirre replied:
I changed that spec, and now the test sets a password "manualy" and then checks that the new password is in the page.
However, I didn't merge two specs together (the one that sets the password manually and this one), because in my opinion both of them check different things: one checks that the password has been correctly set and the user has no problems to log in after that; the other one checks that after saving the password, the admin has the option to print it (looking for the button) and the written password is present only for printing (I couldn't find any way to check the print dialog :confused: )

bertocq replied:
Yes, I expressed myself poorly. I didn't meant to merge both scenarios (the `Manager changes the password by hand (writen by them)`  and `The password is printed`, but to use a manually written password on the `The password is printed` scenario to be able to check it in the printed view (as you correctly did 👏 )

bertocq commented:
@iagirre it's always a good idea to comment on the PR so we can notice requested changes has been applied 👍 we don't cycle through all pending changes PR's refreshing to see if there are changes.

bertocq commented:
The existing user experience for the scenario when the user has an email is confusing:

![screen shot 2018-03-21 at 21 26 44](https://user-images.githubusercontent.com/983242/37735520-9b4956dc-2d4e-11e8-8d57-e64483272198.jpg)

As the "Save" and "Send reset password email" are "final actions" that will leave the current view, but have different colors and positions (bottom seems like final, right like an additional step)

I'll try here (instead of messing with the issue even more) to clarify what's expected **There are two excluding paths/options:**
- Reset by email 
- Reset manually

Similar to what I've said at https://github.com/AyuntamientoMadrid/consul/pull/1363#issuecomment-375081294

1. The "Change user" link should be in the bottom left corner of the grey square, as a red link (check the issue first mockup)
![screen shot 2018-03-21 at 21 35 06](https://user-images.githubusercontent.com/983242/37735929-c164adac-2d4f-11e8-98e7-bd82a58e4f06.jpg)

2. Although the "Reset Password" button was requested as it appears (ignoring a similar feature done by raul) it should be removed from the grey square.
3. A new submenu "Reset Password via Email/Restablecer contraseña via email" should be added under the "Edit user account" menu. And the associated view will only display the current "Send reset password email" button that is already been developed.

![screen shot 2018-03-21 at 21 34 41](https://user-images.githubusercontent.com/983242/37735899-ae5da146-2d4f-11e8-9b64-27c438b6bbfc.jpg)

4. A second new submenu "Reset Password Manually" should be added under the "Edit user account" menu. And the associated view will only display the password reset form that has already been developed. Also when arriving to this form the password field should not be already filled with a value (to avoid confusing the user/manger into thinking that it was the current user password), the same applies to the "Back" button from the "Print password view", as it returns to the form and `12345678` password appears in the input ... instead of the one that we manually set
![screen shot 2018-03-21 at 21 34 37](https://user-images.githubusercontent.com/983242/37735935-c6baeda2-2d4f-11e8-84b1-0a63808b45db.jpg)

5. Random passwords don't need to have strange characters like `gyy5!27e`, its better a large one that an "impossible to read by humans" one https://xkcd.com/936/. If we are forced to generate a random value in JS, just use alphanumeric characters (with upper/lower case letters if you want more combinations) and at least 10 or 12 chars instead of 8

iagirre commented:
I've just made the push to update this PR. I added the submenu with the two options to change the password.

![password01](https://user-images.githubusercontent.com/31625251/37767514-ac7c4d0c-2dca-11e8-9708-6f2475aea698.gif)

I also changed the random password generator so that it doesn't use characters like $ or !, but 12 alphanumeric chars with some capital letters (I maintained the same char group without ambiguous characters like 1, l or I).

**NOTE**: I know that the 'Change user' link is still unchanged, but @raul-fuentes is modifing the same files and he did it, so when both PR are merged the link will appear red and in the bottom of the box (I said this in order to avoid conflics when both PR will be merged, but If I should also change the link, let me know and I will).

**NOTE 2**: In my computer the form to change the password always appears unfilled :thinking: so I couldn't prove if the solution given works correctly. Let me know if not :pray:

bertocq commented:
[1] About the weird password fill behaviour that I reported... 

>  Also when arriving to this form the password field should not be already filled with a value (to avoid confusing the user/manger into thinking that it was the current user password), the same applies to the "Back" button from the "Print password view", as it returns to the form and 12345678 password appears in the input ... instead of the one that we manually set

I realized it was a chrome plugin I have (enabled also for private sessions) that auto-fills password fields 🙈 my bad!

[2] Sorry but spec/features/management/account_spec.rb:15 spec is failing. I guess raul's fix will do the trick https://github.com/AyuntamientoMadrid/consul/pull/1363/files#diff-7f5afe0b6447090de432c81501a61459R14 we can just assume by merging both branches it will be solved 👍 

Thanks for the changes in the menus, I like that there's a dropdown for the new submenus 👏  and also for the neat idea `(I maintained the same char group without ambiguous characters like 1, l or I).` on the random password generation.
